mzLib/FlashLFQ/PEP/PEPValueAnalysisGeneric.cs

```diff
@@ -0,0 +1,670 @@
+                    var threadPredictionEngine = mLContext.Model.CreatePredictionEngine<ChromatographicPeakData, TruePositivePrediction>(threadSpecificTrainedModel);
+
+                    for (int i = range.Item1; i < range.Item2; i++)
+                    {
+                        DonorGroup donor = donors[donorIndices[i]];
+
+                        foreach(ChromatographicPeak peak in donor)
+                        {
+                            ChromatographicPeakData pd = CreateOneChromatographicPeakDataEntry(peak, label: !peak.RandomRt);
+                            var pepValuePrediction = threadPredictionEngine.Predict(pd);
+                            peak.PipPep = 1 - pepValuePrediction.Probability;
+                        }
+                    }
+                });
+        }
+
+        public static string AggregateMetricsForOutput(List<CalibratedBinaryClassificationMetrics> allMetrics)
+        {
+            List<double> accuracy = allMetrics.Select(m => m.Accuracy).ToList();
+            List<double> areaUnderRocCurve = allMetrics.Select(m => m.AreaUnderRocCurve).ToList();
+            List<double> areaUnderPrecisionRecallCurve = allMetrics.Select(m => m.AreaUnderPrecisionRecallCurve).ToList();
+            List<double> F1Score = allMetrics.Select(m => m.F1Score).ToList();
+            List<double> logLoss = allMetrics.Select(m => m.LogLoss).ToList();
+            List<double> logLossReduction = allMetrics.Select(m => m.LogLossReduction).ToList();
+            List<double> positivePrecision = allMetrics.Select(m => m.PositivePrecision).ToList();
+            List<double> positiveRecall = allMetrics.Select(m => m.PositiveRecall).ToList();
+            List<double> negativePrecision = allMetrics.Select(m => m.NegativePrecision).ToList();
+            List<double> negativeRecall = allMetrics.Select(m => m.NegativeRecall).ToList();
+
+            // log-loss can stochastically take on a value of infinity.
+            // correspondingly, log-loss reduction can be negative infinity.
+            // when this happens for one or more of the metrics, it can lead to uninformative numbers.
+            // so, unless they are all infinite, we remove them from the average. If they are all infinite, we report that.
+
+            logLoss.RemoveAll(x => x == Double.PositiveInfinity);
+            logLossReduction.RemoveAll(x => x == Double.NegativeInfinity);
+
+            double logLossAverage = Double.PositiveInfinity;
+            double logLossReductionAverage = Double.NegativeInfinity;
+
+            if ((logLoss != null) && (logLoss.Any()))
+            {
+                logLossAverage = logLoss.Average();
+            }
+
+            if ((logLossReduction != null) && (logLossReduction.Any()))
+            {
+                logLossReductionAverage = logLossReduction.Average();
+            }
+
+            StringBuilder s = new StringBuilder();
```

Comment:
If a gradient boosted decision tree with semi-supervised learning uses the final tree, does it make sense to take the average of them all for output?